pkg/blobstore/local/closable_block_list.go

```diff
@@ -0,0 +1,80 @@
+package local
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/buildbarn/bb-storage/pkg/blobstore/buffer"
+)
+
+// ErrClosedForWriting is an error code to indicate that the BlockList does not
+// accept any more write operations and ongoing write operations will fail.
+var ErrClosedForWriting = errors.New("Location record invalid")
+
+// ClosableBlockList is an implementation of BlockList that can be blocked for
+// writing. This is useful when shutting down to know that no more data is
+// written after the last sync.
+type ClosableBlockList struct {
+	BlockList
+
+	// The lock is used to synchronize closedForWriting with all write methods.
+	lock             sync.RWMutex
+	closedForWriting bool
+}
+
+// NewClosableBlockList provides an implementation of BlockList whose
+// state can be persisted. This makes it possible to preserve the
+// contents of a KeyBlobMap across restarts.
+func NewClosableBlockList(blockList BlockList) *ClosableBlockList {
+	return &ClosableBlockList{
+		BlockList: blockList,
+	}
+}
+
+// CloseForWriting makes this BlockList return ErrClosedForWriting for all
+// future calls to PushBack and Put. This also applies to any ongoing calls
+// that that have not finished when CloseForWriting returns.
+func (bl *ClosableBlockList) CloseForWriting() {
+	bl.lock.Lock()
+	defer bl.lock.Unlock()
+
+	bl.closedForWriting = true
+}
+
+// PushBack appends a new block to the BlockList. The block is obtained
+// by calling into the underlying BlockAllocator.
+func (bl *ClosableBlockList) PushBack() error {
+	bl.lock.RLock()
+	defer bl.lock.RUnlock()
+
+	if bl.closedForWriting {
+		return ErrClosedForWriting
+	}
+	return bl.BlockList.PushBack()
+}
+
+// Put data into a block managed by the BlockList.
+//
+// Even if closed for writing, Put will forward all calls to the backend. It is
+// just in the final call to the BlockListPutFinalizer that ErrClosedForWriting
+// will be returned.
+func (bl *ClosableBlockList) Put(index int, sizeBytes int64) BlockListPutWriter {
+	actualWriter := bl.BlockList.Put(index, sizeBytes)
+
+	return func(b buffer.Buffer) BlockListPutFinalizer {
+		// Copy data into the block without holding any locks.
+		actualFinalizer := actualWriter(b)
+
+		return func() (int64, error) {
```

Comment:
`flatBlobAccess` and `PeriodicSyncer` already ensure that this part of the storage backend is adequately locked. What I mean to say is that `bl.lock` can be eliminated entirely, assuming you set this flag while holding the lock as well.

My recommendation would also be to change the very top up `Put()` to check for `closedForWriting` as well. Otherwise you will end up allocating space and ingesting objects, even though you know you will end up returning an error. Something like this?
```go
if bl.closedForWriting {
	return func(b buffer.Buffer) BlockListPutFinalizer {
		b.Discard()
		return func() (int64, error) {
			return errClosedForWriting
		}
	}
}

Reply:
LocalBlobAccess is a bit of a complex beast, so the locking is also a bit wacky. To clarify it here:
```go
func (bl *ClosableBlockList) Put(index int, sizeBytes int64) BlockListPutWriter {
 	// This is where the exclusive lock is held, so that we can allocate space.
 	return func(b buffer.Buffer) BlockListPutFinalizer {
 		// This is where no locks are held, so that data can be ingested at the client's pace.
 		return func() (int64, error) {
 			// This is where the exclusive lock is held once again, so that we can record that ingestion failed/succeeded.
 		}
 	}
}
```
So comparing against `bl.closedForWriting` inside the middle layer is an illegal chess move here.

Hmmm... Now that I'm taking a step back: in case we have `bl.closedForWriting == true`, we never want to call into `actualFinalizer()`. The reason being that this would:
- Cause the epoch counter to be increased, even though we're terminating/syncing.
- Cause the object to become available for FindMissingBlobs()/Read() to other clients, even though we offer no guarantees it gets preserved.

Unfortunately, you can't get away with _not_ calling `actualFinalizer()`. You have to call it, as PersistentBlockList leaks otherwise. Maybe we should simply integrate this functionality into PersistentBlockList?

Reply:
Comparing `bl.closedForWriting` in the middle layer should still be valid. What I did wrong is that actualWriter must still be called to guarantee no leakage. I'll make the code simpler by removing that case.

I see your point, integrated it.